cmd/juju/application/storage.go

```diff
@@ -0,0 +1,267 @@
+
+    juju application-storage mysql database=,rootfs,1
+
+If no pool is provided, Juju selects the default storage pool from the environment.
+This pool will be recorded as the updated value for the application’s storage.
+
+	juju application-storage mysql database=100G,,1
+
+If no count is provided, Juju uses the minimum count required by the charm. 
+That count will be used when updating the application’s storage.
+
+	juju application-storage mysql database=100G,rootfs,
+`
+)
+
+var applicationStorageConfigBase = config.ConfigCommandBase{
+	Resettable: false,
+}
+
+// NewStorageCommand wraps configCommand with sane application settings.
+func NewStorageCommand() cmd.Command {
+	return modelcmd.Wrap(&storageConfigCommand{configBase: applicationStorageConfigBase})
+}
+
+type storageConfigCommand struct {
+	modelcmd.ModelCommandBase
+	configBase config.ConfigCommandBase
+	out        cmd.Output
+
+	api             StorageConstraintsAPI
+	applicationName string
+}
+
+// Info implements part of the cmd.Command interface.
+func (c *storageConfigCommand) Info() *cmd.Info {
+	info := &cmd.Info{
+		Name:     "application-storage",
+		Args:     "<application-name> [<storage-key>[=<size>,<pool>,<count>]] ...",
+		Purpose:  appStorageConfigSummary,
+		Examples: appStorageConfigExamples,
+		Doc:      appStorageConfigDoc,
+		SeeAlso: []string{
+			"storage",
+			"storage-pools",
+			"add-unit",
+		},
+	}
+
+	return jujucmd.Info(info)
+}
+
+// SetFlags implements part of the cmd.Command interface.
+func (c *storageConfigCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ModelCommandBase.SetFlags(f)
+	c.configBase.SetFlags(f)
+
+	c.out.AddFlags(f, "tabular", map[string]cmd.Formatter{
+		"json":    cmd.FormatJson,
+		"tabular": formatConfigTabular,
+		"yaml":    cmd.FormatYaml,
+	})
+}
+
+// Init implements part of the cmd.Command interface.
+func (c *storageConfigCommand) Init(args []string) error {
+	nArgs := len(args)
+	if nArgs == 0 {
+		return errors.Errorf("no application specified")
+	}
+
+	if !names.IsValidApplication(args[0]) {
+		return errors.Errorf("invalid application name %q", args[0])
+	}
+	c.applicationName = args[0]
+	return c.configBase.Init(args[1:])
+}
+
+// getAPI returns the API. This allows passing in a test configCommandAPI
+// implementation.
+func (c *storageConfigCommand) getAPI() (StorageConstraintsAPI, error) {
+	if c.api != nil {
+		return c.api, nil
+	}
+	api, err := c.NewAPIRoot()
+	if err != nil {
+		return nil, errors.Annotate(err, "opening API connection")
+	}
+	client := application.NewClient(api)
+	return client, nil
+}
+
+// Run implements the meaty part of the cmd.Command interface.
+func (c *storageConfigCommand) Run(ctx *cmd.Context) error {
+	client, err := c.getAPI()
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	for _, action := range c.configBase.Actions {
+		var err error
+		switch action {
+		case config.GetOne:
+			err = c.getConfig(client, ctx)
```

Comment:
it seems really weird to have a ctx as the second argument, thoughts?

Reply:
hi tom do you mean to rearrange the position of the ctx? Not sure what the historical reason was but I was trying to be more consistent with existing code, a key point that was raised in previous PR reviews. Perhaps we were trying to separate cmd context and native go context?